naloge/topo.tex

```diff
@@ -15,6 +15,111 @@
+
+\noindent Trditev lahko dokažemo s premislekom poteka določitev pooznak,
+ko pridemo v algoritmu dfs prvo do u in prvo do v vozlišč.
+Trditev nam omogoča zapis krajšega algoritma, ki bo implementiral dfs.
+Ideja je, da padajoče uredimo vozlišča po njihovih pooznakah,
+po čemer dobimo inverzno topološko ureditev.
+To pa dosežemo tako, da po tem, ko docela preiščemo vozlišče, tega položimo na sklad. \\
+
+\begin{algorithmic}
+\Function{Topo}{V, E}
+	\State $z \gets \length(V)$
+	\State oznaceni $\gets [\False, \dots, \False]$
+	\State toposklad $\gets []$
+	\Function{dfs}{$u$}
+		\State oznaceni[u] = \True
+		\For{$\forall (u, v) \in E$}
+			\If{not oznaceni[$v$]}
+				\State dfs(v)
+			\EndIf
+		\EndFor
+		\State toposklad.\append(u)
+	\EndFunction
+	\For{$\forall u =0, 1, 2, \dots, n - 1$}
+		\If{not oznaceni[$v$]}
+			\State dfs(v)
+		\EndIf
+	\EndFor
+	\State vrni \reverse(toposklad)
+\EndFunction \\
+\end{algorithmic}
+
+\noindent Po klicu topo(Graf), nam ta vrne ureditev vozlišč $[6, 0, 7, 1, 2, 5, 3, 4]$, 
+kar lahko pretvorimo nazaj v začetne oznake $[G, A, H, B, C, F, D, E]$. \pagebreak
+
+\noindent b) Tukaj bomo uporabili rezultat prejšnje naloge in osnoven 
+koncept dinamičnega programiranja.\\
+
+\noindent Naj dist(u) predstavlja najkrajšo razdaljo do vozlišča u.
+Trivialna rešitev je za dist(G), saj je to vozlišče brez vhodnih povezav, torej je najdaljša pot do njega
+enaka 0. Zapišimo rekurzivno formulo: \\
+$dist(u) = min_{v \rightarrow u}(dist(v) + l(v, u))$, \
+kjer l(v, u) določa utež povezave v $\rightarrow$ u \\
+
+\noindent Da bomo imeli vse potrebovane dist(v) definirane pred klicom dist(u) 
+poskrbimo z rezultatom prejšnje naloge, saj pri topološki ureditvi vodijo vse povezave le naprej, 
+torej po premiku na naslednje vozlišče ne potrebujemo vozlišč po, a le ta pred njim,
+katera pa že imamo.\\
+
+\noindent Torej imamo vse kar potrebujemo za izračun najkrajše poti, 
+za beleženje vozlišč skozi katera vodi ta pot pa bomo uporabili seznam predhodnikov, 
+ki bo nakazoval katero vozlišče smo izbrali za predhodnika, torej: \\
+predhodnik(u) $= argmin_{v \rightarrow u}(dist(v) + l(v, u))$ \\
+Na koncu dobimo pot tako da sledimo predhodnikom E, dokler ne pridemo do G.  \\
+
+\begin{algorithmic}
+\Function{najkrajšaPot}{$V, E, s, e$}
+	\State $n \gets \length(V)$
+	\State topo $\gets [6, 0, 7, 1, 2, 5, 3, 4]$
+	\State dist $\gets [\infty, \dots, \infty]$
+	\State dist[s] $\gets 0$
+	\State predhodniki $\gets [None, \dots, None]$
+	\For{$\forall u \in \text{topo}$}
+		\For{$\forall (u, v) \in E$}
+			\If{$\text{dist[v]} \ \textgreater \ \text{dist[u]} + l(u, v)$}
+				\State dist[v] $\gets$ dist[u] + $l(u, v)$
+				\State predhodniki[v] $\gets u$
+			\EndIf
+		\EndFor
+	\EndFor
+	\State pot $\gets [e]$
+	\While{\True}
+		\State predhodnik $\gets$ predhodniki[pot[-1]]
+		\State pot.$\append$(predhodnik)
+		\If{$\text{predhodnik = s ali predhodnik = None}$}
```

Comment:
Se sploh lahko zgodi, da ni predhodnika? To se načeloma lahko zgodi le, če je `s` tudi končno vozlišče (kar se pa lahko takoj preveri). Bolje bi sicer bilo, da negacijo tega pogoja uporabiš pri `\While`.